### PR TITLE
 Bug 1525719: show the list of contributors at the bottom of each page 

### PR DESCRIPTION
**Pull request by davidflanagan**
This PR modifies the document API to include the list of contributors
and the last modified date for each document. On the frontend, it
modifies the React Page component to display that information near
the bottom of each page.

The patch also modifies the DocumentProvider component to add a
localeFromURL property to the document data. This new property is
intented as a single source of truth for the locale that is part
of the currently-displayed URL. Various React components that had
previously been manually extracting the locale from window.location
are modified to use this new localeFromURL property. This relatively
minor change required changes to test files and snapshots, making this
patch appear larger than it actually is.

### CODE REVIEW

**Comment by davidflanagan** [comment]
This patch doesn't really have 4 commits and 29 modified files. It is just based on top of PR #5309. Once that one is reviewed and merged, this PR will be much simpler.

**Comment by schalkneethling** [comment]
Thanks @davidflanagan, let me know once you have merged and rebased.

**Comment by davidflanagan** [comment]
Thanks @schalkneethling. I've merged the other one and rebased this one, so this is ready for review. There was a test failure that I'm hoping was a transient issue, but if there is a real failing test, I'll address that today.

**Comment by davidflanagan** [comment]
I just pushed another commit to fix some python lint and to restart the travis run which seemed hung.

Travis is reporting a test failure for a test that works locally. I'm not sure what's going on...

File: kuma/wiki/tests/test_api.py [hunk 1]
```diff
@@ -48,6 +48,9 @@ def test_doc_api(client, trans_doc):
         'title': 'Root Document',
         'url': '/en-US/docs/Root'
     }]
+    assert data['contributors'] == ['wiki_user']
```
Comment (davidflanagan) [inline kuma/wiki/tests/test_api.py]: This assertion passes when I run the test locally, but is failing on Travis.

Comment (davidflanagan) [reply kuma/wiki/tests/test_api.py]: Hmm. It only passes locally if I run kuma/wiki/tests/test_api.py in isolation. If I do `py.test kuma/wiki` to run all the wiki tests, it fails. So one of the other tests is somehow modifying the trans_doc or wiki_user fixtures

Comment (davidflanagan) [reply kuma/wiki/tests/test_api.py]: I found a fix.

File: kuma/javascript/src/document-provider.jsx [hunk 2]
```diff
@@ -21,7 +21,25 @@ export type DocumentData = {
...
+    // if the user asks for a Spanish translation that does not exist,
+    // they will see "es" in the URL. But the document hat we display
+    // will have an "en-US" locale.  Unlike all of the properties
+    // defined above, this one does not come from the JSON blob of
+    // document data. Instead, we derive this from the URL and add it
+    // the documentData property that we pass to the context provider
+    // below. Even though this property does not exist in the JSON
+    // blobs we load from the backend, we ensure it exists and is
+    // non-null before the data gets used on the frontend, so this
+    // type is defined as if the property is always present.
```
Comment (schalkneethling) [inline kuma/javascript/src/document-provider.jsx]: Should use multiline JSDoc comments

File: kuma/javascript/src/document-provider.jsx [hunk 3]
```diff
@@ -21,7 +21,25 @@ export type DocumentData = {
...
+    }>,
+    contributors: Array<string>,
+    lastModified: string, // An ISO date
+    lastModifiedBy: string,
+
+    // The localeFromURL is the locale that appears to the user in the
+    // location bar. This may be different than the locale of the
+    // document which is what the `locale` property is. For example,
+    // if the user asks for a Spanish translation that does not exist,
+    // they will see "es" in the URL. But the document hat we display
```
Comment (schalkneethling) [inline kuma/javascript/src/document-provider.jsx]:
```suggestion
    // they will see "es" in the URL. But the document that we display
```

File: kuma/javascript/src/document-provider.jsx [hunk 4]
```diff
@@ -21,7 +21,25 @@ export type DocumentData = {
...
+    // will have an "en-US" locale.  Unlike all of the properties
+    // defined above, this one does not come from the JSON blob of
+    // document data. Instead, we derive this from the URL and add it
+    // the documentData property that we pass to the context provider
```
Comment (schalkneethling) [inline kuma/javascript/src/document-provider.jsx]:
```suggestion
    // to the documentData property that we pass to the context provider
```

File: kuma/javascript/src/document-provider.jsx [hunk 5]
```diff
@@ -132,6 +150,12 @@ export default function DocumentProvider(
         });
     }, []);
 
+    // Get the locale displayed in the URL and add that to the data
+    // that we provide.
```
Comment (schalkneethling) [inline kuma/javascript/src/document-provider.jsx]: Should use multiline JSDoc comments

File: kuma/javascript/src/page.jsx [hunk 6]
```diff
@@ -158,12 +185,55 @@ function Quicklinks() {
 function Article() {
     const documentData = useContext(DocumentProvider.context);
     return (
+        // The "text-content" class and "wikiArticle" id are required
+        // because our stylesheets expect them and formatting isn't quite
+        // right without them.
```
Comment (schalkneethling) [inline kuma/javascript/src/page.jsx]: Should use multiline JSDoc comments

**Comment by davidflanagan** [comment]
Thanks @escattone. I think I intentionally punted on the translation stuff, figuring that translators can go to the wiki, and then get the right translation links. But I can certainly adjust that edit URL in a later PR. Thanks for noticing that case.

**Pull request merged**